applications/tests/function_tests/test_cuda_geo_transformer_produce_and_load_coeffs.cpp

```diff
@@ -54,152 +59,237 @@ class GeoTransformerProduceAndLoadCoeffsTest : public ::testing::Test {
         return { dis( gen ), dis( gen ) };
     }
 
-    void randomly_initialize(float* array, int size, int seed) {
+    void randomly_initialize(T* array, int size, int seed) {
         gen.seed( seed );
         std::uniform_real_distribution<> dis(-1.0, 1.0);
 
         for ( int i = 0; i < size; ++i ) {
             array[i] = dis(gen);
         }
-}
-
-void cpu_reference(const float *in, float *out, int x, int y) {
-    const float gain = 6.0f;
-    const float z = sqrtf(3.f) - 2.f;
-    float z1;
-
-    // process lines
-    for (int line = 0; line < y; line++) {
-        float* myLine = out + (line * x);
-
-        // copy input data
-        for (int i = 0; i < x; i++) {
-            myLine[i] = in[i + (line * x)] * gain;
-        }
+    }
 
-        // compute 'sum'
-        float sum = (myLine[0] + powf(z, x)
-            * myLine[x - 1]) * (1.f + z) / z;
-        z1 = z;
-        float z2 = powf(z, 2 * x - 2);
-        float iz = 1.f / z;
-        for (int j = 1; j < (x - 1); ++j) {
-            sum += (z2 + z1) * myLine[j];
-            z1 *= z;
-            z2 *= iz;
+    void cpu_reference(const float *in, float *out, int x, int y) {
+        const float gain = 6.0f;
+        const float z = sqrtf(3.f) - 2.f;
+        float z1;
+
+        // process lines
+        for (int line = 0; line < y; line++) {
+            float* myLine = out + (line * x);
+
+            // copy input data
+            for (int i = 0; i < x; i++) {
+                myLine[i] = in[i + (line * x)] * gain;
+            }
+
+            // compute 'sum'
+            float sum = (myLine[0] + powf(z, x)
```

Comment:
Don't worry too much about this code. As long as it is only for testing, we don't care too much about the performance. Feel free to template this method and use pow for all T

Reply:
Fixed.